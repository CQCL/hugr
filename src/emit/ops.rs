use anyhow::{anyhow, bail, Result};
use hugr::ops::{
    constant::Sum, Call, CallIndirect, Case, Conditional, Const, ExtensionOp, Input, LoadConstant,
    LoadFunction, OpTag, OpTrait, OpType, Output, Tag, Value, CFG,
};
use hugr::{
    hugr::views::SiblingGraph,
    types::{SumType, Type, TypeEnum},
    HugrView, NodeIndex,
};
use inkwell::types::BasicTypeEnum;
use inkwell::{
    builder::Builder,
    values::{BasicValueEnum, CallableValue},
};
use itertools::{zip_eq, Itertools};
use petgraph::visit::Walker;

use crate::{fat::FatExt as _, sum::LLVMSumValue};
use crate::{fat::FatNode, types::LLVMSumType};

use super::{
    deaggregate_call_result,
    func::{EmitFuncContext, RowPromise},
    EmitOp, EmitOpArgs,
};

mod cfg;

struct SumOpEmitter<'c, 'd, H>(&'d mut EmitFuncContext<'c, H>, LLVMSumType<'c>);

impl<'c, 'd, H: HugrView> SumOpEmitter<'c, 'd, H> {
    pub fn new(context: &'d mut EmitFuncContext<'c, H>, st: LLVMSumType<'c>) -> Self {
        Self(context, st)
    }
    pub fn try_new(
        context: &'d mut EmitFuncContext<'c, H>,
        ts: impl IntoIterator<Item = Type>,
    ) -> Result<Self> {
        let llvm_sum_type = context.llvm_sum_type(get_exactly_one_sum_type(ts)?)?;
        Ok(Self::new(context, llvm_sum_type))
    }
}

impl<'c, H: HugrView> EmitOp<'c, Tag, H> for SumOpEmitter<'c, '_, H> {
    fn emit(&mut self, args: EmitOpArgs<'c, Tag, H>) -> Result<()> {
        let builder = self.0.builder();
        args.outputs.finish(
            builder,
            [self.1.build_tag(builder, args.node.tag, args.inputs)?],
        )
    }
}

struct DataflowParentEmitter<'c, 'd, OT, H> {
    context: &'d mut EmitFuncContext<'c, H>,
    node: FatNode<'c, OT, H>,
    inputs: Option<Vec<BasicValueEnum<'c>>>,
    outputs: Option<RowPromise<'c>>,
}

impl<'c, 'd, OT: OpTrait, H: HugrView> DataflowParentEmitter<'c, 'd, OT, H>
where
    for<'a> &'a OpType: TryInto<&'a OT>,
{
    pub fn new(context: &'d mut EmitFuncContext<'c, H>, args: EmitOpArgs<'c, OT, H>) -> Self {
        Self {
            context,
            node: args.node,
            inputs: Some(args.inputs),
            outputs: Some(args.outputs),
        }
    }

    /// safe because we are guarenteed only one input or output node
    fn take_input(&mut self) -> Result<Vec<BasicValueEnum<'c>>> {
        self.inputs
            .take()
            .ok_or(anyhow!("DataflowParentEmitter: Input taken twice"))
    }

    fn take_output(&mut self) -> Result<RowPromise<'c>> {
        self.outputs
            .take()
            .ok_or(anyhow!("DataflowParentEmitter: Output taken twice"))
    }

    pub fn builder(&mut self) -> &Builder<'c> {
        self.context.builder()
    }

    pub fn emit_children(mut self) -> Result<()> {
        use petgraph::visit::Topo;
        let node = self.node;
        if !OpTag::DataflowParent.is_superset(node.tag()) {
            Err(anyhow!("Not a dataflow parent"))?
        };

        let (i, o): (FatNode<Input, H>, FatNode<Output, H>) = node
            .get_io()
            .ok_or(anyhow!("emit_dataflow_parent: no io nodes"))?;
        debug_assert!(i.out_value_types().count() == self.inputs.as_ref().unwrap().len());
        debug_assert!(o.in_value_types().count() == self.outputs.as_ref().unwrap().len());

        let region: SiblingGraph = node.try_new_hierarchy_view().unwrap();
        Topo::new(&region.as_petgraph())
            .iter(&region.as_petgraph())
            .filter(|x| (*x != node.node()))
            .map(|x| node.hugr().fat_optype(x))
            .try_for_each(|node| {
                let inputs_rmb = self.context.node_ins_rmb(node)?;
                let inputs = inputs_rmb.read(self.builder(), [])?;
                let outputs = self.context.node_outs_rmb(node)?.promise();
                self.emit(EmitOpArgs {
                    node,
                    inputs,
                    outputs,
                })
            })
    }
}

impl<'c, OT: OpTrait, H: HugrView> EmitOp<'c, OpType, H> for DataflowParentEmitter<'c, '_, OT, H>
where
    for<'a> &'a OpType: TryInto<&'a OT>,
{
    fn emit(&mut self, args: EmitOpArgs<'c, OpType, H>) -> Result<()> {
        if !OpTag::DataflowChild.is_superset(args.node().tag()) {
            Err(anyhow!("Not a dataflow child"))?
        };

        match args.node().as_ref() {
            OpType::Input(_) => {
                let i = self.take_input()?;
                args.outputs.finish(self.builder(), i)
            }
            OpType::Output(_) => {
                let o = self.take_output()?;
                o.finish(self.builder(), args.inputs)
            }
            _ => emit_optype(self.context, args),
        }
    }
}

struct ConditionalEmitter<'c, 'd, H>(&'d mut EmitFuncContext<'c, H>);

impl<'c, H: HugrView> EmitOp<'c, Conditional, H> for ConditionalEmitter<'c, '_, H> {
    fn emit(
        &mut self,
        EmitOpArgs {
            node,
            inputs,
            outputs,
        }: EmitOpArgs<'c, Conditional, H>,
    ) -> Result<()> {
        let context = &mut self.0;
        let exit_rmb = context
            .new_row_mail_box(node.dataflow_signature().unwrap().output.iter(), "exit_rmb")?;
        let exit_block = context.build_positioned_new_block(
            format!("cond_exit_{}", node.node().index()),
            None,
            |context, bb| {
                let builder = context.builder();
                outputs.finish(builder, exit_rmb.read_vec(builder, [])?)?;
                Ok::<_, anyhow::Error>(bb)
            },
        )?;

        let rmbs_blocks = node
            .children()
            .enumerate()
            .map(|(i, n)| {
                let label = format!("cond_{}_case_{}", node.node().index(), i);
                let node = n.try_into_ot::<Case>().ok_or(anyhow!("not a case node"))?;
                let rmb =
                    context.new_row_mail_box(node.get_io().unwrap().0.types.iter(), &label)?;
                context.build_positioned_new_block(&label, Some(exit_block), |context, bb| {
                    let inputs = rmb.read_vec(context.builder(), [])?;
                    emit_dataflow_parent(
                        context,
                        EmitOpArgs {
                            node,
                            inputs,
                            outputs: exit_rmb.promise(),
                        },
                    )?;
                    context.builder().build_unconditional_branch(exit_block)?;
                    Ok((rmb, bb))
                })
            })
            .collect::<Result<Vec<_>>>()?;

        let sum_type = get_exactly_one_sum_type(node.in_value_types().next().map(|x| x.1))?;
        let sum_input = LLVMSumValue::try_new(inputs[0], context.llvm_sum_type(sum_type)?)?;
        let builder = context.builder();
        sum_input.build_destructure(builder, |builder, tag, mut vs| {
            let (rmb, bb) = &rmbs_blocks[tag];
            vs.extend(&inputs[1..]);
            rmb.write(builder, vs)?;
            builder.build_unconditional_branch(*bb)?;
            Ok(())
        })?;
        builder.position_at_end(exit_block);
        Ok(())
    }
}

fn get_exactly_one_sum_type(ts: impl IntoIterator<Item = Type>) -> Result<SumType> {
    let Some(TypeEnum::Sum(sum_type)) = ts
        .into_iter()
        .map(|t| t.as_type_enum().clone())
        .exactly_one()
        .ok()
    else {
        Err(anyhow!("Not exactly one SumType"))?
    };
    Ok(sum_type)
}

pub fn emit_value<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    v: &Value,
) -> Result<BasicValueEnum<'c>> {
    match v {
        Value::Extension { e } => {
            let exts = context.extensions();
            exts.load_constant(context, e.value())
        }
        Value::Function { .. } => bail!(
            "Value::Function Const nodes are not supported. \
            Ensure you eliminate these from the HUGR before lowering to LLVM. \
            `hugr_llvm::utils::inline_constant_functions` is provided for this purpose."
        ),
        Value::Sum(Sum {
            tag,
            values,
            sum_type,
        }) => {
            let llvm_st = LLVMSumType::try_new(&context.typing_session(), sum_type.clone())?;
            let vs = values
                .iter()
                .map(|x| emit_value(context, x))
                .collect::<Result<Vec<_>>>()?;
            llvm_st.build_tag(context.builder(), *tag, vs)
        }
    }
}

pub(crate) fn emit_dataflow_parent<'c, OT: OpTrait, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, OT, H>,
) -> Result<()>
where
    for<'a> &'a OpType: TryInto<&'a OT>,
{
    DataflowParentEmitter::new(context, args).emit_children()
}

fn emit_tag<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, Tag, H>,
) -> Result<()> {
    SumOpEmitter::try_new(context, args.node.out_value_types().map(|x| x.1))?.emit(args)
}

fn emit_conditional<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, Conditional, H>,
) -> Result<()> {
    ConditionalEmitter(context).emit(args)
}

fn emit_load_constant<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, LoadConstant, H>,
) -> Result<()> {
    let konst_node = args
        .node
        .single_linked_output(0.into())
        .unwrap()
        .0
        .try_into_ot::<Const>()
        .unwrap();
    let r = emit_value(context, konst_node.value())?;
    args.outputs.finish(context.builder(), [r])
}

fn emit_call<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, Call, H>,
) -> Result<()> {
    if !args.node.called_function_type().params().is_empty() {
        return Err(anyhow!("Call of generic function"));
    }
    let (func_node, _) = args
        .node
        .single_linked_output(args.node.called_function_port())
        .unwrap();
    let func = match func_node.as_ref() {
        OpType::FuncDecl(_) => context.get_func_decl(func_node.try_into_ot().unwrap()),
        OpType::FuncDefn(_) => context.get_func_defn(func_node.try_into_ot().unwrap()),
        _ => Err(anyhow!("emit_call: Not a Decl or Defn")),
    };
    let inputs = args.inputs.into_iter().map_into().collect_vec();
    let builder = context.builder();
    let call = builder.build_call(func?, inputs.as_slice(), "")?;
    let call_results = deaggregate_call_result(builder, call, args.outputs.len())?;
    args.outputs.finish(builder, call_results)
}

fn emit_call_indirect<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, CallIndirect, H>,
) -> Result<()> {
    let func_ptr = match args.inputs[0] {
        BasicValueEnum::PointerValue(v) => Ok(v),
        _ => Err(anyhow!("emit_call_indirect: Not a pointer")),
    }?;
    let func =
        CallableValue::try_from(func_ptr).expect("emit_call_indirect: Not a function pointer");
    let inputs = args.inputs.into_iter().skip(1).map_into().collect_vec();
    let builder = context.builder();
    let call = builder.build_call(func, inputs.as_slice(), "")?;
    let call_results = deaggregate_call_result(builder, call, args.outputs.len())?;
    args.outputs.finish(builder, call_results)
}

fn emit_load_function<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, LoadFunction, H>,
) -> Result<()> {
    if !args.node.func_sig.params().is_empty() {
        return Err(anyhow!("Load of generic function"));
    }
    let (func_node, _) = args
        .node
        .single_linked_output(args.node.function_port())
        .unwrap();

    let func = match func_node.as_ref() {
        OpType::FuncDecl(_) => context.get_func_decl(func_node.try_into_ot().unwrap()),
        OpType::FuncDefn(_) => context.get_func_defn(func_node.try_into_ot().unwrap()),
        _ => Err(anyhow!("emit_call: Not a Decl or Defn")),
    }?;
    args.outputs.finish(
        context.builder(),
        [func.as_global_value().as_pointer_value().into()],
    )
}

fn emit_cfg<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, CFG, H>,
) -> Result<()> {
    cfg::CfgEmitter::new(context, args)?.emit_children()
}

fn emit_optype<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, OpType, H>,
) -> Result<()> {
    let node = args.node();
    match node.as_ref() {
        OpType::Tag(ref tag) => emit_tag(context, args.into_ot(tag)),
        OpType::DFG(_) => emit_dataflow_parent(context, args),

        OpType::ExtensionOp(ref co) => {
            let extensions = context.extensions();
            extensions.emit(context, args.into_ot(co))
        }
        OpType::LoadConstant(ref lc) => emit_load_constant(context, args.into_ot(lc)),
        OpType::Call(ref cl) => emit_call(context, args.into_ot(cl)),
        OpType::CallIndirect(ref cl) => emit_call_indirect(context, args.into_ot(cl)),
        OpType::LoadFunction(ref lf) => emit_load_function(context, args.into_ot(lf)),
        OpType::Conditional(ref co) => emit_conditional(context, args.into_ot(co)),
        OpType::CFG(ref cfg) => emit_cfg(context, args.into_ot(cfg)),
        // Const is allowed, but requires no work here. FuncDecl is technically
        // not allowed, but there is no harm in allowing it.
        OpType::Const(_) => Ok(()),
        OpType::FuncDecl(_) => Ok(()),
        OpType::FuncDefn(ref fd) => {
            context.push_todo_func(node.into_ot(fd));
            Ok(())
        }

        _ => Err(anyhow!("Invalid child for Dataflow Parent: {node}")),
    }
}

/// Emit a custom operation with a single input.
///
/// # Arguments
///
/// * `context` - The context in which to emit the operation.
/// * `args` - The arguments to the operation.
/// * `go` - The operation to build the result given a [`Builder`], the input,
///   and an iterator over the expected output types.
pub(crate) fn emit_custom_unary_op<'c, H, F>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, ExtensionOp, H>,
    go: F,
) -> Result<()>
where
    H: HugrView,
    F: FnOnce(
        &mut EmitFuncContext<'c, H>,
        BasicValueEnum<'c>,
        &[BasicTypeEnum<'c>],
    ) -> Result<Vec<BasicValueEnum<'c>>>,
{
    let [inp] = TryInto::<[_; 1]>::try_into(args.inputs).map_err(|v| {
        anyhow!(
            "emit_custom_unary_op: expected exactly one input, got {}",
            v.len()
        )
    })?;
    let out_types = args.outputs.get_types().collect_vec();
    let res = go(context, inp, &out_types)?;
    if res.len() != args.outputs.len()
        || zip_eq(res.iter(), out_types).any(|(a, b)| a.get_type() != b)
    {
        return Err(anyhow!(
            "emit_custom_unary_op: expected outputs of types {:?}, got {:?}",
            args.outputs.get_types().collect_vec(),
            res.iter().map(BasicValueEnum::get_type).collect_vec()
        ));
    }
    args.outputs.finish(context.builder(), res)
}

/// Emit a custom operation with two inputs of the same type.
///
/// # Arguments
///
/// * `context` - The context in which to emit the operation.
/// * `args` - The arguments to the operation.
/// * `go` - The operation to build the result given a [`Builder`], the two
///   inputs, and an iterator over the expected output types.
pub(crate) fn emit_custom_binary_op<'c, H, F>(
    context: &mut EmitFuncContext<'c, H>,
    args: EmitOpArgs<'c, ExtensionOp, H>,
    go: F,
) -> Result<()>
where
    H: HugrView,
    F: FnOnce(
        &mut EmitFuncContext<'c, H>,
        (BasicValueEnum<'c>, BasicValueEnum<'c>),
        &[BasicTypeEnum<'c>],
    ) -> Result<Vec<BasicValueEnum<'c>>>,
{
    let [lhs, rhs] = TryInto::<[_; 2]>::try_into(args.inputs).map_err(|v| {
        anyhow!(
            "emit_custom_binary_op: expected exactly 2 inputs, got {}",
            v.len()
        )
    })?;
    if lhs.get_type() != rhs.get_type() {
        return Err(anyhow!(
            "emit_custom_binary_op: expected inputs of the same type, got {} and {}",
            lhs.get_type(),
            rhs.get_type()
        ));
    }
    let out_types = args.outputs.get_types().collect_vec();
    let res = go(context, (lhs, rhs), &out_types)?;
    if res.len() != out_types.len() || zip_eq(res.iter(), out_types).any(|(a, b)| a.get_type() != b)
    {
        return Err(anyhow!(
            "emit_custom_binary_op: expected outputs of types {:?}, got {:?}",
            args.outputs.get_types().collect_vec(),
            res.iter().map(BasicValueEnum::get_type).collect_vec()
        ));
    }
    args.outputs.finish(context.builder(), res)
}
