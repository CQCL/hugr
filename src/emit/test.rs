use std::iter;

use crate::custom::int::add_int_extensions;
use crate::types::HugrFuncType;
use hugr::builder::DataflowSubContainer;
use hugr::builder::{
    BuildHandle, Container, DFGWrapper, Dataflow, HugrBuilder, ModuleBuilder, SubContainer,
};
use hugr::extension::prelude::BOOL_T;
use hugr::extension::{ExtensionRegistry, EMPTY_REG};
use hugr::ops::constant::CustomConst;
use hugr::ops::handle::FuncID;
use hugr::ops::{CallIndirect, Tag, UnpackTuple, Value};
use hugr::std_extensions::arithmetic::int_ops::{self, INT_OPS_REGISTRY};
use hugr::std_extensions::arithmetic::int_types::ConstInt;
use hugr::types::{Signature, Type, TypeRow};
use hugr::{type_row, Hugr};
use itertools::Itertools;
use rstest::rstest;

use crate::test::*;

#[allow(clippy::upper_case_acronyms)]
pub type DFGW<'a> = DFGWrapper<&'a mut Hugr, BuildHandle<FuncID<true>>>;

pub struct SimpleHugrConfig {
    ins: TypeRow,
    outs: TypeRow,
    extensions: ExtensionRegistry,
}

impl SimpleHugrConfig {
    pub fn new() -> Self {
        Self {
            ins: Default::default(),
            outs: Default::default(),
            extensions: EMPTY_REG,
        }
    }

    pub fn with_ins(mut self, ins: impl Into<TypeRow>) -> Self {
        self.ins = ins.into();
        self
    }

    pub fn with_outs(mut self, outs: impl Into<TypeRow>) -> Self {
        self.outs = outs.into();
        self
    }

    pub fn with_extensions(mut self, extensions: ExtensionRegistry) -> Self {
        self.extensions = extensions;
        self
    }

    pub fn finish(
        self,
        make: impl for<'a> FnOnce(DFGW<'a>) -> <DFGW<'a> as SubContainer>::ContainerHandle,
    ) -> Hugr {
        self.finish_with_exts(|builder, _| make(builder))
    }
    pub fn finish_with_exts(
        self,
        make: impl for<'a> FnOnce(
            DFGW<'a>,
            &ExtensionRegistry,
        ) -> <DFGW<'a> as SubContainer>::ContainerHandle,
    ) -> Hugr {
        let mut mod_b = ModuleBuilder::new();
        let func_b = mod_b
            .define_function("main", HugrFuncType::new(self.ins, self.outs))
            .unwrap();
        make(func_b, &self.extensions);
        mod_b.finish_hugr(&self.extensions).unwrap()
    }
}

impl Default for SimpleHugrConfig {
    fn default() -> Self {
        Self::new()
    }
}

#[macro_export]
macro_rules! check_emission {
    ($hugr: ident, $test_ctx:ident) => {
        let root = $crate::fat::FatExt::fat_root::<hugr::ops::Module>(&$hugr).unwrap();
        let module = $test_ctx
            .get_emit_hugr()
            .emit_module(root)
            .unwrap()
            .finish();

        let mut settings = insta::Settings::clone_current();
        let new_suffix = settings
            .snapshot_suffix()
            .map_or("pre-mem2reg".into(), |x| format!("pre-mem2reg@{x}"));
        settings.set_snapshot_suffix(new_suffix);
        settings.bind(|| insta::assert_snapshot!(module.to_string()));

        module
            .verify()
            .unwrap_or_else(|pp| panic!("Failed to verify module: {pp}"));

        let pb = inkwell::passes::PassManager::create(());
        pb.add_promote_memory_to_register_pass();
        pb.run_on(&module);

        insta::assert_snapshot!(module.to_string());
    };
}

#[rstest]
fn emit_hugr_make_tuple(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_ins(vec![BOOL_T, BOOL_T])
        .with_outs(Type::new_tuple(vec![BOOL_T, BOOL_T]))
        .finish(|mut builder: DFGW| {
            let in_wires = builder.input_wires();
            let r = builder.make_tuple(in_wires).unwrap();
            builder.finish_with_outputs([r]).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_unpack_tuple(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_ins(Type::new_tuple(vec![BOOL_T, BOOL_T]))
        .with_outs(vec![BOOL_T, BOOL_T])
        .finish(|mut builder: DFGW| {
            let unpack = builder
                .add_dataflow_op(
                    UnpackTuple::new(vec![BOOL_T, BOOL_T].into()),
                    builder.input_wires(),
                )
                .unwrap();
            builder.finish_with_outputs(unpack.outputs()).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_tag(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_outs(Type::new_unit_sum(3))
        .finish(|mut builder: DFGW| {
            let tag = builder
                .add_dataflow_op(
                    Tag::new(1, vec![vec![].into(), vec![].into(), vec![].into()]),
                    builder.input_wires(),
                )
                .unwrap();
            builder.finish_with_outputs(tag.outputs()).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_dfg(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_ins(Type::UNIT)
        .with_outs(Type::UNIT)
        .finish(|mut builder: DFGW| {
            let dfg = {
                let b = builder
                    .dfg_builder(HugrFuncType::new_endo(Type::UNIT), builder.input_wires())
                    .unwrap();
                let w = b.input_wires();
                b.finish_with_outputs(w).unwrap()
            };
            builder.finish_with_outputs(dfg.outputs()).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_conditional(llvm_ctx: TestContext) {
    let hugr = {
        let input_v_rows: Vec<TypeRow> = (0..3).map(Type::new_unit_sum).map_into().collect_vec();
        let output_v_rows = {
            let mut r = input_v_rows.clone();
            r.reverse();
            r
        };

        SimpleHugrConfig::new()
            .with_ins(vec![Type::new_sum(input_v_rows.clone()), Type::UNIT])
            .with_outs(vec![Type::new_sum(output_v_rows.clone()), Type::UNIT])
            .finish(|mut builder: DFGW| {
                let cond = {
                    let [sum_input, other_input] = builder.input_wires_arr();
                    let mut cond_b = builder
                        .conditional_builder(
                            (input_v_rows.clone(), sum_input),
                            [(Type::UNIT, other_input)],
                            vec![Type::new_sum(output_v_rows.clone()), Type::UNIT].into(),
                        )
                        .unwrap();
                    for i in 0..3 {
                        let mut case_b = cond_b.case_builder(i).unwrap();
                        let [case_input, other_input] = case_b.input_wires_arr();
                        let tag = case_b
                            .add_dataflow_op(Tag::new(2 - i, output_v_rows.clone()), [case_input])
                            .unwrap();
                        case_b
                            .finish_with_outputs([tag.out_wire(0), other_input])
                            .unwrap();
                    }
                    cond_b.finish_sub_container().unwrap()
                };
                let [o1, o2] = cond.outputs_arr();
                builder.finish_with_outputs([o1, o2]).unwrap()
            })
    };
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_load_constant(mut llvm_ctx: TestContext) {
    llvm_ctx.add_extensions(add_int_extensions);
    let v = Value::tuple([
        Value::unit_sum(2, 4).unwrap(),
        ConstInt::new_s(4, -24).unwrap().into(),
    ]);

    let hugr = SimpleHugrConfig::new()
        .with_outs(v.get_type())
        .with_extensions(INT_OPS_REGISTRY.to_owned())
        .finish(|mut builder: DFGW| {
            let konst = builder.add_load_value(v);
            builder.finish_with_outputs([konst]).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_call(llvm_ctx: TestContext) {
    fn build_recursive(mod_b: &mut ModuleBuilder<Hugr>, name: &str, io: TypeRow) {
        let f_id = mod_b
            .declare(name, HugrFuncType::new_endo(io).into())
            .unwrap();
        let mut func_b = mod_b.define_declaration(&f_id).unwrap();
        let call = func_b
            .call(&f_id, &[], func_b.input_wires(), &EMPTY_REG)
            .unwrap();
        func_b.finish_with_outputs(call.outputs()).unwrap();
    }

    let mut mod_b = ModuleBuilder::new();
    build_recursive(&mut mod_b, "main_void", type_row![]);
    build_recursive(&mut mod_b, "main_unary", type_row![BOOL_T]);
    build_recursive(&mut mod_b, "main_binary", type_row![BOOL_T, BOOL_T]);
    let hugr = mod_b.finish_hugr(&EMPTY_REG).unwrap();
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_call_indirect(llvm_ctx: TestContext) {
    fn build_recursive(mod_b: &mut ModuleBuilder<Hugr>, name: &str, io: TypeRow) {
        let signature = HugrFuncType::new_endo(io);
        let f_id = mod_b.declare(name, signature.clone().into()).unwrap();
        let mut func_b = mod_b.define_declaration(&f_id).unwrap();
        let func = func_b.load_func(&f_id, &[], &EMPTY_REG).unwrap();
        let inputs = iter::once(func).chain(func_b.input_wires());
        let call_indirect = func_b
            .add_dataflow_op(CallIndirect { signature }, inputs)
            .unwrap();
        func_b.finish_with_outputs(call_indirect.outputs()).unwrap();
    }

    let mut mod_b = ModuleBuilder::new();
    build_recursive(&mut mod_b, "main_void", type_row![]);
    build_recursive(&mut mod_b, "main_unary", type_row![BOOL_T]);
    build_recursive(&mut mod_b, "main_binary", type_row![BOOL_T, BOOL_T]);
    let hugr = mod_b.finish_hugr(&EMPTY_REG).unwrap();
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_custom_op(mut llvm_ctx: TestContext) {
    llvm_ctx.add_extensions(add_int_extensions);
    let v1 = ConstInt::new_s(4, -24).unwrap();
    let v2 = ConstInt::new_s(4, 24).unwrap();

    let hugr = SimpleHugrConfig::new()
        .with_outs(v1.get_type())
        .with_extensions(INT_OPS_REGISTRY.to_owned())
        .finish_with_exts(|mut builder: DFGW, ext_reg| {
            let k1 = builder.add_load_value(v1);
            let k2 = builder.add_load_value(v2);
            let ext_op = int_ops::EXTENSION
                .instantiate_extension_op("iadd", [4.into()], ext_reg)
                .unwrap();
            let add = builder.add_dataflow_op(ext_op, [k1, k2]).unwrap();
            builder.finish_with_outputs(add.outputs()).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn get_external_func(llvm_ctx: TestContext) {
    let emc = llvm_ctx.get_emit_module_context();
    let func_type1 = emc.iw_context().i32_type().fn_type(&[], false);
    let func_type2 = emc.iw_context().f64_type().fn_type(&[], false);
    let foo1 = emc.get_extern_func("foo", func_type1).unwrap();
    assert_eq!(foo1.get_name().to_str().unwrap(), "foo");
    let foo2 = emc.get_extern_func("foo", func_type1).unwrap();
    assert_eq!(foo1, foo2);
    assert!(emc.get_extern_func("foo", func_type2).is_err());
}

#[rstest]
fn diverse_module_children(llvm_ctx: TestContext) {
    let hugr = {
        let mut builder = ModuleBuilder::new();
        let _ = {
            let fbuilder = builder
                .define_function("f1", HugrFuncType::new_endo(type_row![]))
                .unwrap();
            fbuilder.finish_sub_container().unwrap()
        };
        let _ = {
            let fbuilder = builder
                .define_function("f2", HugrFuncType::new_endo(type_row![]))
                .unwrap();
            fbuilder.finish_sub_container().unwrap()
        };
        let _ = builder.add_constant(Value::true_val());
        let _ = builder
            .declare("decl", HugrFuncType::new_endo(type_row![]).into())
            .unwrap();
        builder.finish_hugr(&EMPTY_REG).unwrap()
    };
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn diverse_dfg_children(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_outs(BOOL_T)
        .finish(|mut builder: DFGW| {
            let [r] = {
                let mut builder = builder
                    .dfg_builder(HugrFuncType::new(type_row![], BOOL_T), [])
                    .unwrap();
                let konst = builder.add_constant(Value::false_val());
                let func = {
                    let mut builder = builder
                        .define_function("scoped_func", HugrFuncType::new(type_row![], BOOL_T))
                        .unwrap();
                    let w = builder.load_const(&konst);
                    builder.finish_with_outputs([w]).unwrap()
                };
                let [r] = builder
                    .call(func.handle(), &[], [], &EMPTY_REG)
                    .unwrap()
                    .outputs_arr();
                builder.finish_with_outputs([r]).unwrap().outputs_arr()
            };
            builder.finish_with_outputs([r]).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn diverse_cfg_children(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_outs(BOOL_T)
        .finish(|mut builder: DFGW| {
            let [r] = {
                let mut builder = builder.cfg_builder([], type_row![BOOL_T]).unwrap();
                let konst = builder.add_constant(Value::false_val());
                let func = {
                    let mut builder = builder
                        .define_function("scoped_func", HugrFuncType::new(type_row![], BOOL_T))
                        .unwrap();
                    let w = builder.load_const(&konst);
                    builder.finish_with_outputs([w]).unwrap()
                };
                let entry = {
                    let mut builder = builder
                        .entry_builder([type_row![]], type_row![BOOL_T])
                        .unwrap();
                    let control = builder.add_load_value(Value::unary_unit_sum());
                    let [r] = builder
                        .call(func.handle(), &[], [], &EMPTY_REG)
                        .unwrap()
                        .outputs_arr();
                    builder.finish_with_outputs(control, [r]).unwrap()
                };
                let exit = builder.exit_block();
                builder.branch(&entry, 0, &exit).unwrap();
                builder.finish_sub_container().unwrap().outputs_arr()
            };
            builder.finish_with_outputs([r]).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn load_function(llvm_ctx: TestContext) {
    let hugr = {
        let mut builder = ModuleBuilder::new();
        let target_sig = Signature::new_endo(type_row![]);
        let target_func = builder
            .declare("target_func", target_sig.clone().into())
            .unwrap();
        let _ = {
            let mut builder = builder
                .define_function(
                    "main",
                    Signature::new(type_row![], Type::new_function(target_sig)),
                )
                .unwrap();
            let r = builder.load_func(&target_func, &[], &EMPTY_REG).unwrap();
            builder.finish_with_outputs([r]).unwrap()
        };
        builder.finish_hugr(&EMPTY_REG).unwrap()
    };

    check_emission!(hugr, llvm_ctx);
}
