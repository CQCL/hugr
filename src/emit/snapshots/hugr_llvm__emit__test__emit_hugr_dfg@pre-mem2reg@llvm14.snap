---
source: src/emit/test.rs
expression: module.to_string()
---
; ModuleID = 'test_context'
source_filename = "test_context"

define { i32, {} } @_hl.main.1({ i32, {} } %0) {
alloca_block:
  %"0" = alloca { i32, {} }, align 8
  %"2_0" = alloca { i32, {} }, align 8
  %"4_0" = alloca { i32, {} }, align 8
  %"5_0" = alloca { i32, {} }, align 8
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  store { i32, {} } %0, { i32, {} }* %"2_0", align 4
  %"2_01" = load { i32, {} }, { i32, {} }* %"2_0", align 4
  store { i32, {} } %"2_01", { i32, {} }* %"5_0", align 4
  %"5_02" = load { i32, {} }, { i32, {} }* %"5_0", align 4
  store { i32, {} } %"5_02", { i32, {} }* %"4_0", align 4
  %"4_03" = load { i32, {} }, { i32, {} }* %"4_0", align 4
  store { i32, {} } %"4_03", { i32, {} }* %"0", align 4
  %"04" = load { i32, {} }, { i32, {} }* %"0", align 4
  ret { i32, {} } %"04"
}
